### PR TITLE
Use react virtualized

### PR DESCRIPTION
**Pull request by rerissondaniel**
#### What is the purpose of this pull request?

Use [React Virtualized](https://bvaughn.github.io/react-virtualized/#/components/List) to improve the performance of the gallery rendering.

#### What problem is this solving?

Instead of rendering all items, now the gallery only renders the visible items and pre-renders the next items to be shown.

#### How should this be manually tested?
[workspace](https://rerissontst--storecomponents.myvtex.com/clothing)
#### Types of changes

* [X] Bug fix (a non-breaking change which fixes an issue)
* [ ] New feature (a non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Requires change to documentation, which has been updated accordingly.

### CODE REVIEW

File: react/Gallery.js [hunk 1]
```diff
@@ -1,34 +1,104 @@
...
-    'vtex-gallery--two-columns': layoutMode === 'small',
+
+const Gallery = ({
+  products,
+  summary,
+  layoutMode,
+  runtime: { hints: { mobile } },
+  itemWidth }) => {
+  const cache = new CellMeasurerCache({
+    defaultHeight: 305,
```
Comment (luciannojunior) [inline react/Gallery.js]: Fixed value?

File: react/Gallery.js [hunk 2]
```diff
@@ -1,34 +1,104 @@
...
-        )
-      })}
+    <div className="vtex-gallery pa3 bn">
+      <NoSSR onSSR={renderOnServer()}>
+        <WindowScroller>
+          {({ height }) => {
+            return (
+              <AutoSizer disableHeight>
+                {({ width }) => {
+                  const itemsPerRow = (layoutMode === 'small' && mobile) ? 2 : (Math.floor(width / itemWidth) || 1)
```
Comment (luciannojunior) [inline react/Gallery.js]: Extract the 2 into a constant

File: react/Gallery.js [hunk 3]
```diff
@@ -1,34 +1,104 @@
...
+    defaultHeight: 305,
+    fixedWidth: true,
+    keyMapper: () => 1,
   })
 
+  const renderRow = ({ index, key, style, parent, itemsPerRow }) => {
+    const from = index * itemsPerRow
+    const rowItems = products.slice(from, from + itemsPerRow)
+
+    const containerClasses = classNames('Row vtex-gallery__row', {
```
Comment (luciannojunior) [inline react/Gallery.js]: Why this class `Row`?

File: react/Gallery.js (hunk 1 above)
Comment (rerissondaniel) [reply react/Gallery.js]: Its an estimate to the initial unmeasured rows Height, but its not required

File: react/Gallery.js (hunk 3 above)
Comment (rerissondaniel) [reply react/Gallery.js]: I thought it was needed to react virtualized identify the rows, but it isn't, thanks for asking.

File: react/Gallery.js [hunk 4]
```diff
@@ -1,34 +1,105 @@
...
-        )
-      })}
+    <div className="vtex-gallery pa3 bn">
+      <NoSSR onSSR={renderOnServer()}>
+        <WindowScroller>
+          {({ height }) => {
+            return (
+              <AutoSizer disableHeight>
+                {({ width }) => {
+                  const itemsPerRow = (layoutMode === 'small' && mobile) ? TWO_ITEMS : (Math.floor(width / itemWidth) || 1)
```
Comment (luciannojunior) [inline react/Gallery.js]: Add ONE_ITEM. Makes sense

File: react/Gallery.js [hunk 5]
```diff
@@ -1,34 +1,105 @@
...
+  products,
+  summary,
+  layoutMode,
+  runtime: { hints: { mobile } },
+  itemWidth }) => {
+  const TWO_ITEMS = 2
+
+  const cache = new CellMeasurerCache({
+    fixedWidth: true,
+    keyMapper: () => 1,
```
Comment (luciannojunior) [inline react/Gallery.js]: this 1 is for itens too?

File: react/global.css [hunk 6]
```diff
@@ -133,6 +133,16 @@
   margin: 0 auto;
 }
 
+.vtex-gallery__row {
+  display: grid;
+  grid-template-columns: repeat(auto-fill, minmax(280px, 1fr));
```
Comment (luciannojunior) [inline react/global.css]: cant use rem?

Comment (rerissondaniel) [reply react/global.css]: Done, I didn't think could be used here, thanks

File: react/Gallery.js (hunk 5 above)
Comment (rerissondaniel) [reply react/Gallery.js]: Yes, but in this case I think that using named constants doesn't make too much sense. I'll add anyway to maintain consistency.

File: react/Gallery.js [hunk 7]
```diff
@@ -1,34 +1,106 @@
...
+const Gallery = ({
+  products,
+  summary,
+  layoutMode,
+  runtime: { hints: { mobile } },
+  itemWidth }) => {
+  const TWO_ITEMS = 2
+  const ONE_ITEM = 1
+
+  const cache = new CellMeasurerCache({
```
Comment (lucasecdb) [inline react/Gallery.js]: move out of render

File: react/Gallery.js [hunk 8]
```diff
@@ -1,34 +1,106 @@
...
-            />
-          </div>
-        )
-      })}
+    <div className="vtex-gallery pa3 bn">
+      <NoSSR onSSR={renderOnServer()}>
+        <WindowScroller>
+          {({ height }) => {
+            return (
+              <AutoSizer disableHeight>
```
Comment (lucasecdb) [inline react/Gallery.js]: use `Adapt` with this and the `WindowScroller` component

File: react/Gallery.js [hunk 9]
```diff
@@ -1,34 +1,106 @@
...
+                  return (<List
+                    key={layoutMode}
+                    deferredMeasurementCache={cache}
+                    autoHeight
+                    width={width}
+                    height={height}
+                    rowCount={nRows}
+                    rowHeight={cache.rowHeight}
+                    rowRenderer={args => renderRow({ ...args, itemsPerRow })}
+                  />)
```
Comment (lucasecdb) [inline react/Gallery.js]: move the `)` to the line below :pick:

File: react/Gallery.js [hunk 10]
```diff
@@ -1,36 +1,119 @@
 import React from 'react'
-import { PropTypes } from 'prop-types'
 import classNames from 'classnames'
+import { List, WindowScroller, AutoSizer, CellMeasurer, CellMeasurerCache } from 'react-virtualized'
+import { PropTypes } from 'prop-types'
```
Comment (estacioneto) [inline react/Gallery.js]: Are you sure it shouldn't be `import PropTypes from 'prop-types'` ?

Comment (rerissondaniel) [reply react/Gallery.js]: It was already there, thanks for catching it!

File: react/Gallery.js [hunk 11]
```diff
@@ -1,36 +1,119 @@
...
+    <div className={ssrContainer}>
+      {products.map(renderItem)}
     </div>
   )
+
+  return (
+    <NoSSR onSSR={ssrFallBack}>
+      <div className="vtex-gallery pa3 bn">
+        <Adopt
+          key={layoutMode}
```
[the remark this answers is not in the document]
Comment (rerissondaniel) [reply react/Gallery.js]: No, I had added it trying to solve a big, but it wasn't related to that, thanks

**Pull request merged**